components/encryption_udf/encryption_udf_component.cc

```diff
@@ -0,0 +1,796 @@
+        "Maximum RSA key length in bits for create_asymmetric_priv_key()"},
+    // DSA max key length must be <= OPENSSL_DSA_MAX_MODULUS_BITS (10000)
+    // and be a multiple of 64, therefore 9984
+    threshold_definition{
+        1024U, 9984U, "dsa_bits_threshold",
+        "Maximum DSA key length in bits for create_asymmetric_priv_key()"},
+    threshold_definition{
+        1024U, 10000U, "dh_bits_threshold",
+        "Maximum DH key length in bits for create_dh_parameters()"}};
+
+using udf_int_arg_raw_type = mysqlpp::udf_arg_t<INT_RESULT>::value_type;
+bool check_if_bits_in_range(udf_int_arg_raw_type value,
+                            threshold_index_type threshold_index) noexcept {
+  const auto &threshold = thresholds[static_cast<std::size_t>(threshold_index)];
+
+  if (value < static_cast<udf_int_arg_raw_type>(threshold.min)) return false;
+
+  std::size_t max_value = threshold.max;
+
+  static constexpr std::size_t var_buffer_length = 63U;
+  using var_buffer_type = std::array<char, var_buffer_length + 1>;
+  var_buffer_type var_buffer;
+  void *var_buffer_ptr = var_buffer.data();
+  std::size_t var_length = var_buffer_length;
+
+  if (mysql_service_component_sys_variable_register->get_variable(
+          CURRENT_COMPONENT_NAME_STR, threshold.var_name, &var_buffer_ptr,
+          &var_length) == 0) {
+    std::size_t extracted_var_value = 0;
+    if (boost::conversion::try_lexical_convert(
+            static_cast<char *>(var_buffer_ptr), var_length,
+            extracted_var_value))
+      max_value = extracted_var_value;
+  }
+
+  if (value > static_cast<udf_int_arg_raw_type>(max_value)) return false;
+  return true;
+}
+
+// CREATE_ASYMMETRIC_PRIV_KEY(@algorithm, {@key_len|@dh_parameters})
+// This functions generates a private key using the given algorithm
+// (@algorithm) and key length (@key_len) or Diffie-Hellman
+// secret(@dh_parameters), and returns the key as a binary string in PEM format.
+// If key generation fails, the result is NULL.
+class create_asymmetric_priv_key_impl {
+ public:
+  create_asymmetric_priv_key_impl(mysqlpp::udf_context &ctx) {
+    if (ctx.get_number_of_args() != 2)
+      throw std::invalid_argument("Function requires exactly two arguments");
+
+    // result
+    ctx.mark_result_const(false);
+    ctx.mark_result_nullable(true);
+
+    // arg0 - @algorithm
+    ctx.mark_arg_nullable(0, false);
+    ctx.set_arg_type(0, STRING_RESULT);
+
+    // arg1 - @key_len|@dh_parameters
+    ctx.mark_arg_nullable(1, false);
+    ctx.set_arg_type(1, STRING_RESULT);
+  }
+
+  mysqlpp::udf_result_t<STRING_RESULT> calculate(
+      const mysqlpp::udf_context &args);
+};
+
+mysqlpp::udf_result_t<STRING_RESULT> create_asymmetric_priv_key_impl::calculate(
+    const mysqlpp::udf_context &ctx) {
+  auto algorithm = ctx.get_arg<STRING_RESULT>(0);
+  if (algorithm.data() == nullptr)
+    throw std::invalid_argument("Algorithm cannot be NULL");
+  if (algorithm != "RSA" && algorithm != "DSA" && algorithm != "DH")
```

Comment:
Shouldn't these be case insensitive?

Reply:
Agree, even better, I think we should define a conversion function from a string to an enum
enum class algorithm_type { unknowm rsa, dsa, dh }
algorithm_type get_algorith_by_name(const std::string& name) noexcept {...}

Reply:
Introduced `get_algorithm_id_by_label()` and `get_and_validate_algorithm_id_by_label()` functions